### ISSUE TITLE
Unable to find an entry point named 'GetDpiForMonitor' in DLL 'Shcore.dll'.

### ISSUE THREAD

**Issue by Kizaemon** [dnGrep/dnGrep#372]
When installing 2.9.174 x64 on Microsoft Windows [Version 6.2.9200]

```
<Error time="10-16-2019 06:23:50.646">
  <Message>Unable to find an entry point named 'GetDpiForMonitor' in DLL 'Shcore.dll'.</Message>
  <Exception>System.EntryPointNotFoundException
   at DockFloat.NativeMethods.GetDpiForMonitor(IntPtr hmonitor, MonitorDpiType dpiType, UInt32&amp; dpiX, UInt32&amp; dpiY)
   at DockFloat.Extensions.ToDevicePixels(Screen screen, Rect rect)
   at DockFloat.Extensions.IsOnScreen(Rect windowBounds)
   at dnGREP.WPF.MainForm.&lt;&gt;c__DisplayClass3_0.&lt;.ctor&gt;b__0(Object s, RoutedEventArgs e)
   at System.Windows.EventRoute.InvokeHandlersImpl(Object source, RoutedEventArgs args, Boolean reRaised)
   at System.Windows.UIElement.RaiseEventImpl(DependencyObject sender, RoutedEventArgs args)
   at System.Windows.BroadcastEventHelper.BroadcastEvent(DependencyObject root, RoutedEvent routedEvent)
   at System.Windows.BroadcastEventHelper.BroadcastLoadedEvent(Object root)
   at MS.Internal.LoadedOrUnloadedOperation.DoWork()
   at System.Windows.Media.MediaContext.FireLoadedPendingCallbacks()
   at System.Windows.Media.MediaContext.FireInvokeOnRenderCallbacks()
   at System.Windows.Media.MediaContext.RenderMessageHandlerCore(Object resizedCompositionTarget)
   at System.Windows.Media.MediaContext.RenderMessageHandler(Object resizedCompositionTarget)
   at System.Windows.Interop.HwndTarget.OnResize()
   at System.Windows.Interop.HwndTarget.HandleMessage(WindowMessage msg, IntPtr wparam, IntPtr lparam)
   at System.Windows.Interop.HwndSource.HwndTargetFilterMessage(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean&amp; handled)
   at MS.Win32.HwndWrapper.WndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean&amp; handled)
   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(Object o)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)</Exception>
</Error>
```

**Comment by doug24** [comment]
I assumed this API would be safe since it was added to Windows 6 years ago and in all supported versions of Windows:  GetDpiForMonitor requires version 6.3.9600  - Windows 8.1 or Windows Server 2012 R2.

Microsoft no longer supports these [old versions of Windows](https://en.wikipedia.org/wiki/List_of_Microsoft_Windows_versions), although companies can purchase extended support for some of them.

I've tried hard to keep the minimum requirements for dnGrep low, but at some point the application does need to move on and stay up to date.   If there is enough demand for this, I could add a fallback to some other method call.   

How many are still running Windows 7, 8, 8.1, Server 2012, or Server 2012 R2?

**Comment by JVimes** [comment]
In general, I don't think dnGrep has enough developer resources to support old versions of Windows. But if there's an easy way to bypass the error and keep the app usable (even if it gracefully degrades), that sounds okay.

**Comment by doug24** [comment]
I have a fix ready for the next release that falls back to assuming all monitors have the same dpi scaling, which should be a reasonable assumption for older systems.

**Comment by oivanenko** [comment]
Hi!

> I have a fix ready for the next release that falls back to assuming all monitors have the same dpi scaling, which should be a reasonable assumption for older systems.

Seems like it is sill not working for me. I have same error in 2.9.181.0 version, additionally installer is never ending and stuck at some gauge position.

![изображение](https://user-images.githubusercontent.com/201648/68181598-08943f80-ffa9-11e9-92e2-78a639eb1dc4.png)

![изображение](https://user-images.githubusercontent.com/201648/68181619-1ea20000-ffa9-11e9-8913-4aeeed506105.png)

![изображение](https://user-images.githubusercontent.com/201648/68181708-5c068d80-ffa9-11e9-9257-6feda61b3781.png)

Windows 7 64-bit ver. 6.1.7601

**Comment by doug24** [comment]
I don't have any way to test this since I don't have anything running old versions of Windows...

That's actually a different error:  'DllNotFoundException' instead of 'EntryPointNotFoundException'.  

I'll have to add another exception handler or change to using a version number check.

**Comment by oivanenko** [comment]
> That's actually a different error: 'DllNotFoundException' instead of 'EntryPointNotFoundException'.
My bad :(
 
> I'll have to add another exception handler or change to using a version number check.

Sounds great! Thank you for your time and work!

**Comment by doug24** [comment]
Should be fixed in 2.9.184, but I don't have any Windows 7, 8, 8.1, Server 2012, or Server 2012 R2 to test on.

**Comment by Kizaemon** [comment]
Works very well now, tested on Windows Version 6.2.9200.

Thank you for supporting of older systems.
Sometimes stuck with older systems for various reasons.
